lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,569 @@
+<?php
+/***
+ * Scans through an HTML document to find specific tags, then
+ * transforms those tags by adding, removing, or updating the
+ * values of the HTML attributes within that tag (opener).
+ *
+ * Does not fully parse HTML or _recurse_ into the HTML structure
+ * Instead this scans linearly through a document and only parses
+ * the HTML tag openers.
+ *
+ * @TODO:
+ *  - All the return values from `scan()` are written as if
+ *    they are simple string values but they need to be the
+ *    structure of the matches.
+ *      - [ start, end ] of tag opener structure in source
+ *      - [ start,  end ] offset of tag/attribute name in source
+ *      - comparable( $name ) of the tag/attribute name
+ *      - $value of the attribute value
+ *
+ *  - Write a comprehensive test suite for various unexpected
+ *    quirks and attribute names. Unfortunately we need to have
+ *    some fairly robust support because it's not difficult to
+ *    create misleading HTML through the block editor. For
+ *    example, as can happen with a number of plaintext inputs.
+ */
+
+
+/**
+ * Processes an input HTML document by applying a specified set
+ * of patches to that input. Tokenizes HTML but does not fully
+ * parse the input document.
+ */
+class WP_HTML_Processor {
+	/**
+	 * Describes the constraints imposed on the HTML tag we're trying to match.
+	 *
+	 * @see WP_Tag_Find_Descriptor::parse()
+	 *
+	 * @var WP_Tag_Find_Descriptor
+	 */
+	public $descriptor;
+
+
+	/**
+	 * Scans the HTML document, keeping track of its own state.
+	 *
+	 * @var WP_HTML_Scanner
+	 */
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
+		$this->scanner->reset();
+		$this->descriptor = WP_Tag_Find_Descriptor::parse( $query );
+	}
+
+
+	public function apply() {
+		// find tag
+		// apply modifications
+		// construct output
+
+		// @TODO: This is a stub; implement the real apply() behavior.
+		return $this->scanner->document;
+	}
+
+
+	public function __toString() {
+		return $this->apply();
+	}
+}
+
+
+/**
+ * Describes the search conditions for finding a given tag in an HTML document.
+ */
+class WP_Tag_Find_Descriptor {
+	/**
+	 * We're looking for an HTML tag of this name, up to the comparable
+	 * equivalence of those names (lower-cased, Unicode-normalized, etc...).
+	 * If we're looking for "any tag" then this property will be `null`.
+	 *
+	 * h1...h6 are special since they are variations of the same base tag.
```

Comment:
Can you use the list of HTML tags instead of a single value for headings? Example:

`h1,h2,h3,h4`

Reply:
considered it yes, though I would have gone `->find( [ 'h1', 'h2', 'h3', 'h4' ] )`

so far this has been more of an exercise in discovering how _little_ functionality we can expose while being useful. we discussed the idea of a list and wondered if there was any other HTML tag like the heading tags, where multiple tag names all basically relate to the same tag with a different attribute value. if there are not other cases like that I'd almost prefer not exposing a list until we know for sure we need one.

I can see where it could be useful to say "give me the `div` or the `p` with the `most-important` class, but then again if we limit our search for that class maybe that's enough and we don't have to check the tag.

what do you think?

performance wise I don't think it should be that much of an impact _with this current design_ of scanning the full HTML source, since we have to traverse every tag linearly as we come to it (vs. being able to search for something like `<div`, which we can't do without opening up vulnerabilities for common cases of HTML produced by the editor).

still, every bit of flexibility makes me nervous unless I can see a clear need for it.

Reply:
There is also `ul` vs `ol` where those tags are used for different styling.

Less related to that would be all the tags with semantic meaning used instead of `div` like `header`, `main`, `section`, `article`, `aside` or `footer`.

It's also hard to tell how this API would be used with custom blocks and how flexible people expect it to be. If the primary goal is to allow changing the root element of the block content, then using the `most-important` class would be sufficient in most of cases.

Reply:
Yeah I think _most_ cases can be handled appropriately with a class. The outermost and first tag can be found with `$processor->find( [] )` as well so that's not an obstacle.

I considered `ol` and `ul` but thought they generally are different enough in _meaning_ that they don't overlap the same way `h1` through `h6` do (for example, there was plenty of debate about replacing `h1` through `h6` with just `h` and ignoring the numbered heading level when constructing document structure, something the specification allowed/mandated? but was never adopted). The same goes for the other elements; other tags have related semantics, which is often captured in the placement of an element into its content category, but these are distinct elements with distinct semantics. The confusion over whether a certain sub-sub-section should be `h3` or `h4` and the mess of properly styling them isn't something we face with `article` and `main` and `aside` I believe; I think they are qualitatively different, and again, nobody is confused by whether we should use `ol` or `ul` to represent an ordered list, because they are related but different (Gutenberg possibly is the exception in merging both kinds of lists into one block and using an attribute to distinguish them).

So maybe eventually we expand this but I'd like to try and see how far we can get without it. I haven't supported the `h` exception yet and maybe there's a case to make that we never do.

Reply:
@gziolo the point about alternative markup depending on the attributes is spot-on! The list block does this, for example:

```js
const tagName = ordered ? 'ol' : 'ul';
```

Also, the heading block may use either of the h-tags.

Already at this point we'd need to either introduce special tokens for "ol/ul" and "h1-6", or simply allow an array of tags as an argument. If we do the former, we'll introduce a "silent" dependency between the HTML parser and the block library. I vote for the latter even if it opens other possibilities.

In addition, the following blocks have a `tagName` attribute:

* template-part
* comment
* group
* query

Reply:
when thinking about `render_block` though we don't need special-casing for the list or the `tagName` blocks because that information is readily available to us via the JSON attributes of the block.

```php
function render_query_block( $html, $block ) {
    $p = new WP_Tag_Thing( $html );
    return $p->find( [ 'tag_name' => $block['attrs']['tagName'] ] )
             ->set_attribute( 'data-post-id', $post_id );
}

function render_list_block( $html, $block ) {
    $p = new WP_Tag_Thing( $html );
    return $p->find( [ 'tag_name' => $block['attrs']['ordered'] ? 'ol' : 'ul' ] )
             ->add_class( 'todo' );
}
```

Reply:
Good point @dmsnell. I was about to counter that it would be more convenient to just query for ['ol','ul'], but it's not very future-proof. I like matching the exact tag as you proposed. In fact, we could do the same thing in the heading block since `level` is a block attribute as well.

Reply:
I don't think we necessarily need to support an option to pass an array of tags. @dmsnell made a great point that when rendering the block, attributes help to make an exact match for which tag should be processed. 

I left my initial question came from the fact that I didn't fully understand why there is a need for the special `h` tag. It feels like we can live without both `h` and `[ 'h1', 'h2' ]` expressions for now.

Aside. If someone really needs to process all six heading types they can do it with:

```php
$p = new WP_Tag_Thing( $html );
for ( $i = 1; $i <=6; $i++ ) {
    while ( $p->find( [ 'tag_name' => 'h' + $i  ] ) ) {
        $p->add_class( 'todo' );
    }
}
```

Reply:
@gziolo not quite, this is "pull" parser which means it only moves forward. `find` means "Find me the next tag, starting where I finished last".  In other words, this would only process the `div`:

```php
$p = new WP_Tag_Thing('<span><div>');
$p->find('div')->add_class('todo');
$p->find('span')->add_class('todo');
```

While this would process both tags:

```php
$p = new WP_Tag_Thing('<span><div>');
$p->find('span')->add_class('todo');
$p->find('div')->add_class('todo');
```

I'd like to name that method `find_next`, `next_tag`, `next`, or something to that effect.

Reply:
How about the following:
```js
for ( $i = 1; $i <= 6; $i++ ) {
    $p = new WP_Tag_Thing( $html );
    while ( $p->find( [ 'tag_name' => 'h' + $i  ] ) ) {
        $p->add_class( 'todo' );
    }
    $html = (string) $p;
}
```

Reply:
> why there is a need for the special h tag

it's not felt yet, but I anticipated it because the `h` tags are unique in HTML in their purpose and realization. I don't think any other tags work the same way they do where their semantics are contextual (based on the surrounding content vs. based on the tag itself).

> If someone really needs to process all six heading types they can do it with:

This isn't actually possible with the interface we've exposed here because we only go once and linearly through the document, so for example, if we look for `h1` and there are none in the document, we will never find an `h2` if it's there because we would have already scanned to the end.

To find _any_ of the `h` tags we'd have to re-process the entire document from the beginning until we find one.

```php
for ( $i = 1; $i <= 6; $i++ ) {
	$p = new WP_Tag_Walker( $html );
	if ( $p->find( [ 'tag_name' => "h{$i}" ] ) ) {
		return $p->add_class( 'todo' );
	}
}
```

for small blocks this may not be too bad, but note that if we want chain these together we'd have to sequence them.

```php
$output = $html;
foreach ( array( 'ol', 'li' ) as $tag_name ) {
	$p = new WP_Html_Tag_Modifier( $output );
	while ( $p->find( array( 'tag_name' => $tag_name ) ) ) {
		$p->find( array( 'tag_name' => 'li' ) );
		$p->add_class( 'first-item' );
	}
	$output .= $p;
}
return $output.
```